### PR TITLE
Add JSON-java decoder and encoder

### PR DESCRIPTION
**Pull request by radio-rogal**
Added module for encoding and decoding using JSON-java.

Implements #1385

### CODE REVIEW

File: json/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+JSON-java Codec
+===================
+
+This module adds support for encoding and decoding JSON via [JSON-java][].
```
Comment (velo) [inline json/README.md:4]: could you please add the link to https://www.json.org/json-en.html

File: json/pom.xml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>io.github.openfeign</groupId>
+    <artifactId>parent</artifactId>
+    <version>-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>feign-json</artifactId>
```
Comment (velo) [inline json/pom.xml:26]: @kdavisk6 any opinion about the package/module names?  Just calling it "json" triggers something on the back of my mind, not that I have a better suggestion.

File: json/src/main/java/feign/json/JsonEncoder.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ *   github.create("openfeign", "feign", contributor);
+ * </pre>
+ */
+public class JsonEncoder implements Encoder {
+
+  @Override
+  public void encode(Object object, Type bodyType, RequestTemplate template)
+      throws EncodeException {
+    if (JSONArray.class == bodyType || JSONObject.class == bodyType) {
+      if (object == null)
```
Comment (velo) [inline json/src/main/java/feign/json/JsonEncoder.java:57]: could you move this outside the first `if` block?!

No reason to not "work" on any nulls.

File: json/src/main/java/feign/json/JsonEncoder.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ *   contributor.put("contributions", 0);
+ *   github.create("openfeign", "feign", contributor);
+ * </pre>
+ */
+public class JsonEncoder implements Encoder {
+
+  @Override
+  public void encode(Object object, Type bodyType, RequestTemplate template)
+      throws EncodeException {
+    if (JSONArray.class == bodyType || JSONObject.class == bodyType) {
```
Comment (velo) [inline json/src/main/java/feign/json/JsonEncoder.java:56]: wonder if we should instead use `object instanceof JSONArray || ... JSONObject`

also, it's possible to anyone to extend `JSONArray`. Worse case scenario would need to use `isAssignableFrom`

My preference is for instance of on the encoder

File: json/src/main/java/feign/json/JsonDecoder.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        throw (IOException) jsonException.getCause();
+      }
+      throw new DecodeException(response.status(), jsonException.getMessage(), response.request(),
+          jsonException);
+    }
+  }
+
+  private Object decodeJSON(Response response, Type type, Reader reader) {
+    JSONTokener tokenizer = new JSONTokener(reader);
+    if (JSONObject.class == type)
```
Comment (velo) [inline json/src/main/java/feign/json/JsonDecoder.java:77]: Anyone can extends JSONObject would be better to use `isAssignableFrom` here to make sure any subtypes would also work.

Please include inheritance on the codec test too.

**Review by velo — CHANGES_REQUESTED** [review]
Excellent work @radio-rogal , I asked for a few changes, lemme know how it goes.

File: json/pom.xml (hunk 2 above)
Comment (kdavisk6) [reply json/pom.xml:26]: Since the implementation here is `org.json`, I'm good with it.

**Comment by radio-rogal** [comment]
@velo done

File: json/README.md (hunk 1 above)
Comment (radio-rogal) [reply json/README.md]: +

File: json/pom.xml (hunk 2 above)
Comment (radio-rogal) [reply json/pom.xml]: I have renamed it to `feign-org.json` and module to `org.json`.

File: json/src/main/java/feign/json/JsonEncoder.java (hunk 3 above)
Comment (radio-rogal) [reply json/src/main/java/feign/json/JsonEncoder.java]: done

File: json/src/main/java/feign/json/JsonEncoder.java (hunk 4 above)
Comment (radio-rogal) [reply json/src/main/java/feign/json/JsonEncoder.java]: replaced with instanceof

File: json/src/main/java/feign/json/JsonDecoder.java (hunk 5 above)
Comment (radio-rogal) [reply json/src/main/java/feign/json/JsonDecoder.java]: +

File: json/pom.xml (hunk 2 above)
Comment (velo) [reply json/pom.xml]: Ow, no, feign-json was fine

Comment (radio-rogal) [reply json/pom.xml:26]: Sorry, misunderstood

I has returned back feign-json as artifact name and json as directory name.

Comment (radio-rogal) [reply json/pom.xml:26]: How about json-java or org.json as folder/mofule name? The artifact name is still feign-json.

I also did not like that the name is too common, but the artifact of the library, like its name (json-java), is also too general and I could not choose another name.

File: json/src/main/java/feign/json/JsonDecoder.java (hunk 5 above)
Comment (velo) [reply json/src/main/java/feign/json/JsonDecoder.java:77]: @radio-rogal can we get this done? As far I can tell, is the only thing preventing me from merging. I just merged master into your branch, so, it's almost ready

Comment (radio-rogal) [reply json/src/main/java/feign/json/JsonDecoder.java]: @velo Sorry for late response, looks like I have missed this point.

I have changed it in 3d14686

**Review by velo — APPROVED** [review]

**Review by kdavisk6 — APPROVED** [review]

**Pull request merged**